slides/advanced-01-feature-engineering.qmd

```diff
@@ -0,0 +1,783 @@
+
+Feature hashing ([_FES_](https://bookdown.org/max/FES/encoding-predictors-with-many-categories.html),[_SMLTAR_](https://smltar.com/mlregression.html#case-study-feature-hashing), [_TMwR_](https://www.tmwr.org/categorical.html#feature-hashing)):  
+
+- uses the character value of the levels 
+- converts them to integer hash value
+- uses the integer to assign them to a specific indicator column. 
+
+## Feature Hashing
+
+Suppose we want to use 32 indicator variables for `agent`. 
+
+For a agent with value "`Max_Kuhn`", a hashing function converts it to an integer (say `r strtoi(substr(rlang::hash("Max_Kuhn"), 26, 32), 16)`). 
+
+To assign it to one of the 32 columns, we would use modular arithmetic to assign it to a column: 
+
+```{r}
+#| label: hash
+
+# For "Max_Kuhn" but a '1' in column: 
+210397726 %% 32
+```
+
+[Hash functions](https://www.metamorphosite.com/one-way-hash-encryption-sha1-data-software) are used in cryptology and are meant to _emulate_ randomness. 
+
+
+## Feature Hashing Pros and Cons
+
+- The procedure will automatically work on new values of the predictors.
+- The is no real logic behind which factor levels are combined. 
+- It is fast. 
+- We don't know how many columns to add (more in the next section).
+- Some columns may have all zeros. 
+- "Signed" hashes add a sign to help avoid aliasing. 
+- If a indicator column is important to the model, we can't easily determine why. 
+
+
+## Feature Hashing in recipes
+
+The textrecipes package has a step that can be added to the recipe: 
+
+```{r}
+#| label: hash-rec
+#| code-line-numbers: "6-8|"
+library(textrecipes)
+
+hash_rec <-
+  recipe(avg_price_per_room ~ ., data = hotel_tr) %>%
+  step_YeoJohnson(lead_time) %>%
+  # Defaults to 32 signed indicator columns
+  step_dummy_hash(agent) %>%
+  step_dummy_hash(company) %>%
+  # Regular indicators for the others
+  step_dummy(all_nominal_predictors()) %>% 
+  step_zv(all_predictors())
+
+hotel_hash_wflow <-
+  hotel_lm_wflow %>%
+  update_recipe(hash_rec)
+```
+
+
+## Feature Hashing in recipes
+
+```{r}
+#| label: hash-res
+#| cache: true
+
+hotel_hash_res <-
+  hotel_hash_wflow %>%
+  fit_resamples(hotel_rs, control = ctrl)
+
+collect_metrics(hotel_hash_res)
+```
+
+About the same performance but now we can handle new values. 
+
+
+## Debugging a recipe
+
+- Typically, you will want to use a workflow to estimate and apply a recipe.
+
+. . .
+
+- If you have an error and need to debug your recipe, the original recipe object (e.g. `hash_rec`) can be estimated manually with a function called `prep()`. It is analogous to `fit()`. See [TMwR section 16.4](https://www.tmwr.org/dimensionality.html#recipe-functions)
+
+. . .
+
+- Another function (`bake()`) is analogous to `predict()`, and gives you the processed data back.
+
+. . .
+
+- The `tidy()` function can be used to get specific results from the recipe.
+
+## Example
+
+```{r}
+#| label: prep-tidy-bake
+#| eval: false
```

Comment:
Is this intentional?

Reply:
Yes. It is like that in the previous versions (I think because there is a lot of potential output).